pkg/plugincmdtree/plugins_cache.go

```diff
@@ -0,0 +1,257 @@
+// Copyright 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package plugincmdtree
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+	"gopkg.in/yaml.v2"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/cli"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/common"
+)
+
+const pluginsCommandTreeDir = "plugins_command_tree"
+
+type pluginKey struct {
+	PluginName string `yaml:"pluginName" json:"pluginName"`
+	Target     string `yaml:"target" json:"target"`
+	Version    string `yaml:"version" json:"version"`
+}
+
+type pluginCommandTree struct {
+	CommandTree map[pluginKey]*CommandNode `yaml:"commandTree" json:"commandTree"`
+}
+
+func getPluginsCommandTreeCacheDir() string {
+	// NOTE: TEST_CUSTOM_PLUGIN_COMMAND_TREE_CACHE_DIR is only for test purpose
+	customCommandTreeCacheDirForTest := os.Getenv("TEST_CUSTOM_PLUGIN_COMMAND_TREE_CACHE_DIR")
+	if customCommandTreeCacheDirForTest != "" {
+		return customCommandTreeCacheDirForTest
+	}
+	return filepath.Join(common.DefaultCacheDir, pluginsCommandTreeDir)
+}
+func GetPluginsCommandTreeCachePath() string {
+	return filepath.Join(getPluginsCommandTreeCacheDir(), "command_tree.yaml")
+}
+
+func getPluginsDocsCachePath() string {
+	return filepath.Join(getPluginsCommandTreeCacheDir(), "docs")
+}
+
+type cacheImpl struct {
+	pluginCommands      *pluginCommandTree
+	pluginDocsGenerator func(plugin *cli.PluginInfo) error
+}
+
+func NewCache() (Cache, error) {
+	pct, err := getPluginCommandTree()
+	if err != nil {
+		return nil, err
+	}
+	return &cacheImpl{
+		pluginCommands:      pct,
+		pluginDocsGenerator: generatePluginDocs,
+	}, nil
+}
+
+func (c *cacheImpl) GetTree(plugin *cli.PluginInfo) (*CommandNode, error) {
+	// This is just a safety net to generate the command tree if we missed/failed to generate the plugin command tree during plugin install
+	// If the plugin command tree exists, then ConstructAndAddTree is a no-op
+	if err := c.ConstructAndAddTree(plugin); err != nil {
+		return nil, err
+	}
+
+	pkey := pluginKey{
```

Comment:
Something I learned today :-)
That said, I believe in #391  there was some discussion of using the plugininfo's plugin path as a uniq identifier. Would that work here as well for some consistency?

Reply:
Yes, it works (since CLI installation path includes plugin name,version,target and SHA, it gives enough information to debug if there is any undesired behavior), I updated the PR to use plugininfo's plugin path as a uniq identifier.